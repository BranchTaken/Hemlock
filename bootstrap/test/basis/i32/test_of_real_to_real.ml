open! Basis.Rudiments
open! Basis
open I32

let test () =
  let rec test_rs rs = begin
    match rs with
    | [] -> ()
    | r :: rs' -> begin
        let x = of_real r in
        File.Fmt.stdout
        |> Fmt.fmt "of_real "
        |> Real.fmt ~alt:true ~base:Fmt.Hex r
        |> Fmt.fmt " -> "
        |> fmt ~alt:true ~zpad:true ~width:8L ~base:Fmt.Hex ~pretty:true x
        |> Fmt.fmt "; to_real -> "
        |> Real.fmt ~alt:true ~base:Fmt.Hex (to_real x)
        |> Fmt.fmt "\n"
        |> ignore;
        test_rs rs'
      end
  end in
  let rs = [
    -1.;
    0.;
    0x1.1p-1;
    1.;

    0x1.f_ffff_ffff_ffffp48;
    0x1.f_ffff_ffff_ffffp52;
    0x1.f_ffff_ffff_ffffp56;

    0x1.f_ffff_ffff_ffffp61;

    0x1p30;
    0x1p31;
  ] in
  test_rs rs;
  File.Fmt.stdout |> Fmt.fmt "\n" |> ignore;
  let rec test_xs xs = begin
    match xs with
    | [] -> ()
    | x :: xs' -> begin
        let r = to_real x in
        File.Fmt.stdout
        |> Fmt.fmt "to_real "
        |> fmt ~alt:true ~zpad:true ~width:8L ~base:Fmt.Hex ~pretty:true x
        |> Fmt.fmt " -> "
        |> Real.fmt ~alt:true ~base:Fmt.Hex r
        |> Fmt.fmt "; of_real -> "
        |> fmt ~alt:true ~zpad:true ~width:8L ~base:Fmt.Hex ~pretty:true (of_real r)
        |> Fmt.fmt "\n"
        |> ignore;
        test_xs xs'
      end
  end in
  let xs = [
    neg_one;
    zero;
    one;
    max_value;
  ] in
  test_xs xs

let _ = test ()
