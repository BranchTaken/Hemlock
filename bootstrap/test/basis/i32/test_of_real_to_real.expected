of_real -0x1. -> -0x0000_0001i32; to_real -> -0x1.
of_real 0x0. -> 0x0000_0000i32; to_real -> 0x0.
of_real 0x0.88 -> 0x0000_0000i32; to_real -> 0x0.
of_real 0x1. -> 0x0000_0001i32; to_real -> 0x1.
of_real 0x1_ffff_ffff_ffff.f -> 0x7fff_ffffi32; to_real -> 0x7fff_ffff.
of_real 0x1f_ffff_ffff_ffff. -> 0x7fff_ffffi32; to_real -> 0x7fff_ffff.
of_real 0x1ff_ffff_ffff_fff0. -> 0x7fff_fff0i32; to_real -> 0x7fff_fff0.
of_real 0x3fff_ffff_ffff_fe00. -> 0x7fff_fe00i32; to_real -> 0x7fff_fe00.
of_real 0x1p30 -> 0x4000_0000i32; to_real -> 0x1p30
of_real 0x1p31 -> 0x0000_0000i32; to_real -> 0x0.

to_real -0x0000_0001i32 -> -0x1.; of_real -> -0x0000_0001i32
to_real 0x0000_0000i32 -> 0x0.; of_real -> 0x0000_0000i32
to_real 0x0000_0001i32 -> 0x1.; of_real -> 0x0000_0001i32
to_real 0x7fff_ffffi32 -> 0x7fff_ffff.; of_real -> 0x7fff_ffffi32
